### ISSUE TITLE
No syntax highlighting for files opened with "Open repository version"

### ISSUE THREAD

**Issue by holgerbrandl** [vsch/idea-multimarkdown#588]
Any other languages tested (just 2) is providing syntax highlighting for files opened via "Open repository version"
![image](https://user-images.githubusercontent.com/200952/39633282-cd77a380-4fb7-11e8-97b6-c3d9ae150104.png)

![image](https://user-images.githubusercontent.com/200952/39633366-0afe7878-4fb8-11e8-8c8c-cc6f165e2a2e.png)

MN does not, but it would be cool if it would. Preview works fine.

**Comment by vsch** [comment]
@holgerbrandl, never even tried this feature.

I will take a look to see why it is not syntax highlighting.

**Comment by vsch** [comment]
@holgerbrandl, just as I suspected. Repository files don't do annotation phase, so if you change your syntax highlighter to "Lexer" then it will work. I'll take a look to see if I can tell the difference between these files and regular ones (I can use the ReadOnly status) and switch to lexer automatically.

<img width="1265" alt="screen shot 2018-05-04 at 10 40 55 am" src="https://user-images.githubusercontent.com/10299086/39634130-f250f966-4f87-11e8-9573-4d9b1e44ee09.png">

**Comment by holgerbrandl** [comment]
Thanks, works great.

Do I loose any functionality when using "Lexer". I'm curious because since it's a project wide setting, it will also affect all the other markdown documents in my project.

**Comment by vsch** [comment]
@holgerbrandl, both do the same job but at different phases. Lexer is applied as soon as the file is parsed, annotator much later. Annotator has better typing response since it is done after a short break in UI activity.

Some injected language syntax highlighting will appear a bit different between the two because of the different phases.

**Comment by vsch** [comment]
@holgerbrandl, I checked and the API makes it difficult to make the detection automatic since when it requests a syntax highlighter it does not provide any information about the file for which it is requested.

I will have to see if I can change the syntax highlighter after the file is opened so that Read-Only files use lexer by default since annotator does not work for them.

**Comment by holgerbrandl** [comment]
Thanks for looking into it. It's also just a minor detail, so feel welcome to drop it if it turns out to be complicated.

**Comment by vsch** [comment]
@holgerbrandl, I found no way to use lexer for one set of files and annotator for another. I would recommend using the lexer but the IDE has some optimizations which make the lexer syntax highlighting not work well for an actively edited file.

Markdown does not lend itself to parsing a snippet without including the proper context. For example:

```markdown
* list item
  * sub item  
    1. sub item
    2. another sub item
```

I have found lexer sometimes will syntax highlight 1. and 2. lines as indented code which can only happen if lexing is done on these lines without including the parent context. These lines by themselves do look like indented code (4 spaces) only when the parent list items are included do these look like ordered sub-items.

If you are willing to live with such unpleasantries during editing, using lexer over annotator will work.

As an option I can add two Actions: switch to lexer and switch to annotator. Then you can assign a keyboard shortcut so you can switch modes without going into options. I can even bring the option out to a toolbar button so you can switch it without a shortcut. I can probably make the current editors reparse and use the new setting without having to re-open them. So you could switch mode after opening the diff view.

I did just that in the last EAP for gutter icons. Now turning off some gutter icon options will reduce the gutter margin without having to close and re-open the file.

**Comment by holgerbrandl** [comment]
No need to add any options (for me at least). It's a rather hidden feature anyway, and I just reported it because I thought it might be an easy to fix minor bug.

**Comment by vsch** [comment]
@holgerbrandl, actually the actions to change the setting is an easy solution which would address my annoyance at this issue. I noticed it before but did not give it much thought until you brought it up and I looked into it further. Now it bothers me. 😄

**Comment by vsch** [comment]
@holgerbrandl, just have to state the problem in the correct terms. No way to tell for which file or editor the lexer is being created but!!!

I played a bit trying to get the highlighter to update and realized that the lexer is created when editor is created so it won't change because the file never changes.

A break point and a few minutes later, I realized that I can change the highlighter after the editor is created and force it to Lexer. Monitors the editor factory and 100ms after an editor is created if it is read only and markdown will replace its highlighter with a lexer highlighter.

I kludged it because did not want to go deeper into the implmentation so I set lexer highlighter type, create a new highlighter, then reset it back to annotator. May have consequences if many read only and non-read only editors are opened in fast succession but for now its all good.

EAP update with a fix for the issue is available.

[Installing EAP Plugin Version](https://github.com/vsch/idea-multimarkdown/wiki/EAP-Update-Channel)
